### PR TITLE
Rethink 'Raw' Quil

### PR DESCRIPTION
**Pull request by kilimanjaro**
Previously `PARSE-QUIL-INTO-RAW-PROGRAM` produced a `PARSED-PROGRAM` object, which in particular had no notion of where or in what order its various definitions originated. This is fine if the presence of multiple conflicting definitions results in undefined behavior, but e.g. it might be nice to have the option for more fine-grained control in name resolution. Further, the quilt proposal allows (and even encourages) multiple calibrations for a given gate, and we need to be able to support that adequately. After discussions with @ecpeterson and @stylewarning I've opted to shuffle things around a bit.

This PR changes `PARSE-QUIL-INTO-RAW-PROGRAM` to produce a list of ast objects in their source order. `PROCESS-INCLUDES` and `RESOLVE-APPLICATIONS` have been downgraded to ordinary functions (i.e. they are no longer registered transforms), which are invoked by `PARSE-QUIL`.  The other transforms are still valid and work with `PARSED-PROGRAM` objects, as before. 

When includes are processed, we maintain a table mapping definition signatures to (name . definition) pairs. When a collision is observed, this prompts a signal to be raised. The policy for handling these should be explicit. Right now the only handler I have in mind (seen in `PARSE-QUIL`) is for memory declarations (which are treated rather strictly in the parser), but this could naturally be extended for the other definition types. Note that quilt will also add `DEFWAVEFORM` and `DEFCAL`, so the flexibility here may prove useful.

There are a few other changes that go with this. Tokens now get a filename slot.

### CODE REVIEW

File: src/parser.lisp [hunk 1]
```diff
@@ -1636,26 +1659,17 @@ result of BODY, and the (possibly null) list of remaining lines.
...
-          (extract-code-sections parsed-program)
-        (make-instance 'parsed-program
-                       :gate-definitions gate-defs
-                       :circuit-definitions circ-defs
-                       :memory-definitions memory-defs
-                       :executable-code (coerce exec-code 'simple-vector))))))
+  (let* ((*memory-region-names* nil)
+         (tok-lines (tokenize string)))
+    (loop :with parsed-program := nil
+          :until (null tok-lines) :do
```
Comment (stylewarning) [inline src/parser.lisp]:
```suggestion
          :until (endp tok-lines) :do
```

File: tests/cfg-tests.lisp [hunk 2]
```diff
@@ -24,7 +24,7 @@
   (is (string= (format nil "H 0~%CNOT 0 1~%")
                (with-output-to-string (out)
                  (quil::print-code-list (quil:parsed-program-executable-code
-                                         (quil::parse-quil-into-raw-program (format nil "H 0~%CNOT 0 1"))) out)))))
+                                         (quil::parse-quil (format nil "H 0~%CNOT 0 1"))) out)))))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]:
```suggestion
                                         (quil::parse-quil "H 0;CNOT 0 1")) out)))))
```

File: tests/cfg-tests.lisp [hunk 3]
```diff
@@ -44,7 +44,7 @@
 
 (deftest test-reconstitute-unconditional-jump ()
   "Test that the correct program is parsed from a CFG with one edge"
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "H 0~%JUMP @NEXT~%LABEL @NEXT~%X 1")))
+  (let* ((p (quil::parse-quil (format nil "H 0~%JUMP @NEXT~%LABEL @NEXT~%X 1")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]:
```suggestion
  (let* ((p (quil::parse-quil "H 0;JUMP @NEXT;LABEL @NEXT;X 1"))
```

File: tests/cfg-tests.lisp [hunk 4]
```diff
@@ -55,7 +55,7 @@
     (is (search blk-2 string-result))))
 
 (deftest test-reconstitute-conditional-jump ()
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "DECLARE ro BIT~%H 0~%JUMP-UNLESS @NEXT ro[0]~%LABEL @NEXT~%Z 0")))
+  (let* ((p (quil::parse-quil (format nil "DECLARE ro BIT~%H 0~%JUMP-UNLESS @NEXT ro[0]~%LABEL @NEXT~%Z 0")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]:
```suggestion
  (let* ((p (quil::parse-quil "DECLARE ro BIT;H 0;JUMP-UNLESS @NEXT ro[0];LABEL @NEXT;Z 0"))
```

File: tests/cfg-tests.lisp [hunk 5]
```diff
@@ -79,7 +79,7 @@
 
 (deftest test-remove-block ()
   "Tests that removing a block from the CFG preserves the validity of the CFG."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "X 0~%JUMP @END~%LABEL @END")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: ok im tired of writing `;` for all of these

File: tests/misc-tests.lisp [hunk 6]
```diff
@@ -21,14 +21,6 @@
     (test-it t '((2 2)) '(2 2))
     (test-it t '((2) (1 1) (2 2 2) (1 1 1 1)) '(2 1 1 2 2 2 1 1 1 1))))
 
-(deftest test-code-splicing ()
```
Comment (stylewarning) [inline tests/misc-tests.lisp]: rip

File: tests/parser-tests.lisp [hunk 7]
```diff
@@ -201,12 +201,12 @@ TEST 0 1 2"))
...
-(deftest test-process-include-declaration-collision ()
-  (uiop:with-temporary-file (:stream stream :pathname path)
-    (format stream "DECLARE foo BIT~@
-                    MEASURE 0 foo")
-    (force-output stream)
-    (let* ((test-quil (format nil "DECLARE foo BIT~@
-                                   H 0~@
-                                   INCLUDE \"~A\"" path)))
-      (signals quil-parse-error (parse-quil test-quil)))))
+;; (deftest test-process-include-declaration-collision ()
```
Comment (stylewarning) [inline tests/parser-tests.lisp]: why is this disabled? we should test something like this

**Comment by stylewarning** [comment]
this is another erik davis hit PR

File: src/analysis/process-includes.lisp [hunk 8]
```diff
@@ -4,62 +4,70 @@
 
 (in-package #:cl-quil)
 
-(define-transform process-includes (process-includes)
-  "This transform expands (and therefore eliminates) INCLUDE directives."
-  )
+(defun definition-signature (instr)
+  "If instr is a definition or memory declaration, returns a signature used for
```
[the remark this answers is not in the document]
Comment (kilimanjaro) [reply src/analysis/process-includes.lisp]: eagle eyes

File: src/analysis/process-includes.lisp [hunk 9]
```diff
@@ -4,62 +4,70 @@
...
-                      (handle-include pp incl pos originating-directory))))))
-      (expand-all-includes pp))))
+(defun process-includes (raw-quil &optional originating-file)
+  "Recursively process all INCLUDE instructions in the list RAW-QUIL. The
+result is a new sequence with the included quil instructions spliced in."
+  (let* (;; We track the expanded list of instructions, and build it from its tail out
+         (expanded (cons nil nil))
+         (expanded-tail expanded)
+         ;; We track seen files to check for cycles in the include graph
+         (seen-files (make-hash-table :test 'equalp))
```
[the remark this answers is not in the document]
Comment (kilimanjaro) [reply src/analysis/process-includes.lisp]: I've been living a lie

File: tests/cfg-tests.lisp (hunk 5 above)
Comment (kilimanjaro) [reply tests/cfg-tests.lisp]: https://www.fileformat.info/info/unicode/char/037e/index.htm

File: src/analysis/process-includes.lisp (hunk 9 above)
[the remark this answers is not in the document]
Comment (appleby) [reply src/analysis/process-includes.lisp]: What if your filesystem isn't case sensitive?

```
$ uname
Darwin
$ touch foo.txt FOO.txt
$ ls *.txt
foo.txt
```

File: src/analysis/process-includes.lisp [hunk 10]
```diff
@@ -4,62 +4,70 @@
...
+               (let ((file (funcall *resolve-include-pathname* (include-pathname incl)))
+                     (originating-dir (if (null originating-file)
+                                          (pathname (uiop:getcwd))
+                                          (uiop:pathname-directory-pathname
+                                           originating-file))))
+                 (when (uiop:relative-pathname-p file)
+                   (setf file (merge-pathnames file originating-dir)))
+                 (unless (uiop:file-exists-p file)
+                   (error "Could not include ~S because it does not exist." file))
+                 (when (gethash file seen-files)
```
Comment (ecpeterson) [inline src/analysis/process-includes.lisp]: I'd believe this if `seen-files` were a stack that you popped off when you finished including a given file. As a hash table, won't this break if my code has `INCLUDE QFFT.quil ; blah blah ; INCLUDE QFFT.quil` in it?

Comment (kilimanjaro) [reply src/analysis/process-includes.lisp]: Oh, very true

Comment (ecpeterson) [reply src/analysis/process-includes.lisp]: (i had this problem when writing this region of `operator-bind`: https://github.com/rigetti/quilc/blob/master/src/define-compiler.lisp#L891 , & there are only like 50 distinct programming situations)

File: src/analysis/process-includes.lisp (hunk 9 above)
[the remark this answers is not in the document]
Comment (kilimanjaro) [reply src/analysis/process-includes.lisp]: I'm canonicalizing with `TRUENAME` now

File: src/analysis/process-includes.lisp [hunk 11]
```diff
@@ -4,62 +4,70 @@
...
+               (cond ((endp instrs)
+                      (cdr expanded))
+                     ((typep (car instrs) 'include)
+                      (handle-include (car instrs) originating-file)
+                      (expand-all-includes (cdr instrs) originating-file))
+                     (t
+                      (let ((instr (car instrs)))
+                        (a:when-let ((signature (definition-signature instr)))
+                          ;; check for conflicts
+                          (with-simple-restart (continue "Continue with ambiguous definition.")
```
[the remark this answers is not in the document]
Comment (kilimanjaro) [reply src/analysis/process-includes.lisp]: Opting for the former.

File: src/analysis/resolve-applications.lisp [hunk 12]
```diff
@@ -4,11 +4,78 @@
 
 (in-package #:cl-quil)
 
-(define-transform resolve-applications (resolve-applications)
-  "A transform which converts a parsed program with unresolved applications to one where the applications have been resolved into gate or circuit applications.")
-
+;;; Used solely for application resolution to detect if we are resolving things inside of a circuit.
```
Comment (stylewarning) [inline src/analysis/resolve-applications.lisp]: M-q this comment

File: src/parser.lisp [hunk 13]
```diff
@@ -25,10 +25,13 @@
 
 (defvar *line-start-position* nil)
 (defvar *line-number* nil)
+(defvar *current-file* nil
+  "The pathname for the file being currently parsed.")
 
-(defstruct (token (:constructor tok (type &optional payload (line *line-number*))))
+(defstruct (token (:constructor tok (type &optional payload (line *line-number*) (pathname *current-file*))))
   "A lexical token."
   (line nil :type (or null (integer 1)))
```
Comment (stylewarning) [inline src/parser.lisp]: can these first three things be `:read-only t`? Can _all_ of them be read-only?

File: src/parser.lisp [hunk 14]
```diff
@@ -303,6 +309,15 @@ the immediately preceding line."
 (defvar *formal-arguments-allowed* nil
   "Dynamic variable to control whether formal parameters and arguments are allowed in the current parsing context.")
 
+(defmacro tag-with-context (tok &body body)
+  "Execute BODY and set the token context of the first resulting value to TOK."
+  (a:with-gensyms (results ast-obj)
+    `(let* ((,results (multiple-value-list
```
Comment (stylewarning) [inline src/parser.lisp]: not excited about M-V-L and V-L just to tag things,

also who has a token context? seems prone to error

Comment (stylewarning) [reply src/parser.lisp]: possible alternative: pass context to the actual parsing functions that make the objects

File: src/parser.lisp [hunk 15]
```diff
@@ -328,7 +343,8 @@ the immediately preceding line."
 
        (let ((*definitions-allowed* nil)
              (*formal-arguments-allowed* t))
-         (parse-circuit-definition tok-lines)))
+         (tag-with-context tok
+             (parse-circuit-definition tok-lines))))
```
Comment (stylewarning) [inline src/parser.lisp]: example: pass a `:context` argument to this function

File: tests/cfg-tests.lisp [hunk 16]
```diff
@@ -141,7 +141,7 @@
 
 (deftest test-block-fusion-multiple-contraction-unique-code ()
   "Tests the operation of block fusion when the CFG has multiple edges that can be contracted, where each block has disimilar code."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "LABEL @START~%X 0~%JUMP @MID~%LABEL @MID~%Y 0~%JUMP @END~%LABEL @END~%Z 0")))
+  (let* ((p (quil::parse-quil (format nil "LABEL @START;X 0;JUMP @MID;LABEL @MID;Y 0;JUMP @END;LABEL @END;Z 0")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format-nil

File: tests/cfg-tests.lisp [hunk 17]
```diff
@@ -132,7 +132,7 @@
 
 (deftest test-block-fusion-simple-contraction ()
   "Tests the operation of block fusion when the CFG has a single edge that can be contracted."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "X 0~%JUMP @NEXT~%LABEL @NEXT~%JUMP @ANOTHER~%LABEL @ANOTHER~%JUMP @END~%LABEL @END")))
+  (let* ((p (quil::parse-quil (format nil "X 0;JUMP @NEXT;LABEL @NEXT;JUMP @ANOTHER;LABEL @ANOTHER;JUMP @END;LABEL @END")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 18]
```diff
@@ -119,7 +119,7 @@
 
 (deftest test-block-fusion-maximally-simplified ()
   "Tests the operation of block fusion when the CGF is already maximally simplified."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "X 0")))
+  (let* ((p (quil::parse-quil (format nil "X 0")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 19]
```diff
@@ -89,14 +89,14 @@
 
 (deftest test-dce-none ()
   "Tests the operation of dead code elimination when the cfg is already optimal."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "X 0~%JUMP @END~%LABEL @END")))
+  (let* ((p (quil::parse-quil (format nil "X 0;JUMP @END;LABEL @END")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 20]
```diff
@@ -79,7 +79,7 @@
 
 (deftest test-remove-block ()
   "Tests that removing a block from the CFG preserves the validity of the CFG."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "X 0~%JUMP @END~%LABEL @END")))
+  (let* ((p (quil::parse-quil (format nil "X 0;JUMP @END;LABEL @END")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 21]
```diff
@@ -89,14 +89,14 @@
...
          (cfg (quil::program-cfg p :dce t))
          (blocks-result (quil::cfg-blocks cfg)))
     (is (= 2 (length blocks-result)))))
 
 (deftest test-dce-simple-extras ()
   "Tests the operation of dead code elimination when presented with a simple graph with extra labels."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "X 0~%JUMP @END~%LABEL @END~%Y 0~%JUMP @ANOTHER~%LABEL @ANOTHER~%Z 0~%JUMP @END~%LABEL @END")))
+  (let* ((p (quil::parse-quil (format nil "X 0;JUMP @END;LABEL @END;Y 0;JUMP @ANOTHER;LABEL @ANOTHER;Z 0;JUMP @END;LABEL @END")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 22]
```diff
@@ -105,7 +105,7 @@
 
 (deftest test-dce-dead-loop ()
   "Tests the operation of dead code elimination when the cfg has a dead loop."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "JUMP @L~%LABEL @D~%X 0~%JUMP @D~%LABEL @L~%H 0")))
+  (let* ((p (quil::parse-quil (format nil "JUMP @L;LABEL @D;X 0;JUMP @D;LABEL @L;H 0")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 23]
```diff
@@ -159,7 +159,7 @@
 
 (deftest test-block-fusion-empty-single-unconditional-self-loop ()
   "Tests the operation of block fusion when the CFG has an unconditional self-loop with a single empty block."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "LABEL @START~%H 0~%JUMP @LOOPER~%LABEL @LOOPER~%JUMP @START")))
+  (let* ((p (quil::parse-quil (format nil "LABEL @START;H 0;JUMP @LOOPER;LABEL @LOOPER;JUMP @START")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 24]
```diff
@@ -169,7 +169,7 @@
 
 (deftest test-block-fusion-nonempty-conditional-multiple-self-loop ()
   "Tests the operation of block fusion when the CFG has a long self loop with code inside."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "DECLARE ro BIT~%LABEL @START~%H 0~%JUMP-WHEN @LOOPER ro[0]~%JUMP @END~%LABEL @LOOPER~%JUMP @LOOPER2~%LABEL @LOOPER2~%X 0~%JUMP @LOOPER3~%LABEL @LOOPER3~%JUMP @START~%LABEL @END")))
+  (let* ((p (quil::parse-quil (format nil "DECLARE ro BIT;LABEL @START;H 0;JUMP-WHEN @LOOPER ro[0];JUMP @END;LABEL @LOOPER;JUMP @LOOPER2;LABEL @LOOPER2;X 0;JUMP @LOOPER3;LABEL @LOOPER3;JUMP @START;LABEL @END")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 25]
```diff
@@ -179,7 +179,7 @@
 
 (deftest test-block-fusion-empty-conditional-self-loop ()
   "Tests the operaton of block fusion when the CFG has an empty self loop with a conditional control."
-  (let* ((p (quil::parse-quil-into-raw-program (format nil "DECLARE ro BIT~%LABEL @START~%H 0~%JUMP-WHEN @LOOPER ro[0]~%JUMP @END~%LABEL @LOOPER~%JUMP @LOOPER2~%LABEL @LOOPER2~%JUMP @LOOPER3~%LABEL @LOOPER3~%JUMP @START~%LABEL @END")))
+  (let* ((p (quil::parse-quil (format nil "DECLARE ro BIT;LABEL @START;H 0;JUMP-WHEN @LOOPER ro[0];JUMP @END;LABEL @LOOPER;JUMP @LOOPER2;LABEL @LOOPER2;JUMP @LOOPER3;LABEL @LOOPER3;JUMP @START;LABEL @END")))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: tests/cfg-tests.lisp [hunk 26]
```diff
@@ -192,5 +192,5 @@
 
 This is a regression test for https://github.com/rigetti/quilc/issues/244"
   (dolist (jump-instr '("JUMP-WHEN" "JUMP-UNLESS"))
-    (let ((p (quil::parse-quil-into-raw-program (format nil "DECLARE ro BIT~%LABEL @START~%~A @START ro[0]" jump-instr))))
+    (let ((p (quil::parse-quil (format nil "DECLARE ro BIT;LABEL @START;~A @START ro[0]" jump-instr))))
```
Comment (stylewarning) [inline tests/cfg-tests.lisp]: remove format nil

File: src/ast.lisp [hunk 27]
```diff
@@ -292,7 +292,13 @@ If no exit rewiring is found, return NIL."
 
 ;;;;;;;;;;;;;;;;;;;; Gate and Circuit Definitions ;;;;;;;;;;;;;;;;;;;;
 
-(defclass gate-definition ()
+(defclass context-token-mixin ()
```
[the remark this answers is not in the document]
Comment (kilimanjaro) [reply src/ast.lisp]: Fair call, I'm downgrading from `lexical-context` class to a `lexical-context` type and a generic on ast objs

File: src/cl-quil.lisp [hunk 28]
```diff
@@ -5,16 +5,37 @@
...
-(defun parse-quil (string &key originating-file (transforms *standard-post-process-transforms*))
-  "Parse and process the Quil string STRING, which originated from the file ORIGINATING-FILE. Transforms in TRANSFORMS are applied in-order to the processed Quil string."
-  (let ((pp (parse-quil-into-raw-program string)))
-    (dolist (xform transforms pp)
-      (case xform
-        (process-includes (setf pp (transform xform pp originating-file)))
-        (otherwise        (setf pp (transform xform pp)))))))
+(defun parse-quil (string &key originating-file
+                            (transforms *standard-post-process-transforms*)
+                            (ambiguous-definition-handler #'continue))
```
[the remark this answers is not in the document]
Comment (kilimanjaro) [reply src/cl-quil.lisp]: Switching from `signal` to `cerror` and using `continue` as the default handler.

**Pull request merged**